File: src/game/action.ts
import type { ActivePair, Field, Move, Rotation } from './types';
import { reachableTargets } from './reachability';

export const ACTION_COUNT = 22;

export function moveToActionIndex(move: Move): number {
  const { axisCol, rotation } = move;
  if (!Number.isInteger(axisCol) || axisCol < 0 || axisCol > 5) {
    throw new Error(`invalid axisCol: ${axisCol}`);
  }
  if (rotation === 0) return axisCol;
  if (rotation === 2) return 6 + axisCol;
  if (rotation === 1) {
    if (axisCol < 0 || axisCol > 4) throw new Error(`rot=1 axisCol out of range: ${axisCol}`);
    return 12 + axisCol;
  }
  if (rotation === 3) {
    if (axisCol < 1 || axisCol > 5) throw new Error(`rot=3 axisCol out of range: ${axisCol}`);
    return 17 + axisCol - 1;
  }
  throw new Error(`invalid rotation: ${String(rotation)}`);
}

export function actionIndexToMove(index: number): Move {
  if (!Number.isInteger(index) || index < 0 || index >= ACTION_COUNT) {
    throw new Error(`invalid action index: ${index}`);
  }
  if (index < 6) return { axisCol: index, rotation: 0 as Rotation };
  if (index < 12) return { axisCol: index - 6, rotation: 2 as Rotation };
  if (index < 17) return { axisCol: index - 12, rotation: 1 as Rotation };
  return { axisCol: index - 17 + 1, rotation: 3 as Rotation };
}

export function legalActionMask(field: Field, start: ActivePair): Uint8Array {
  const reachable = reachableTargets(field, start);
  const mask = new Uint8Array(ACTION_COUNT);
  for (let i = 0; i < ACTION_COUNT; i++) {
    const { axisCol, rotation } = actionIndexToMove(i);
    if (reachable.has(`${axisCol}-${rotation}`)) mask[i] = 1;
  }
  return mask;
}

File: python/puyo_train/action.py
from __future__ import annotations

ACTION_COUNT = 22


def move_to_action_index(axis_col: int, rotation: int) -> int:
    if not isinstance(axis_col, int) or axis_col < 0 or axis_col > 5:
        raise ValueError(f"invalid axis_col: {axis_col}")
    if rotation == 0:
        return axis_col
    if rotation == 2:
        return 6 + axis_col
    if rotation == 1:
        if axis_col < 0 or axis_col > 4:
            raise ValueError(f"rot=1 axis_col out of range: {axis_col}")
        return 12 + axis_col
    if rotation == 3:
        if axis_col < 1 or axis_col > 5:
            raise ValueError(f"rot=3 axis_col out of range: {axis_col}")
        return 17 + axis_col - 1
    raise ValueError(f"invalid rotation: {rotation}")


def action_index_to_move(index: int) -> tuple[int, int]:
    if not isinstance(index, int) or index < 0 or index >= ACTION_COUNT:
        raise ValueError(f"invalid action index: {index}")
    if index < 6:
        return (index, 0)
    if index < 12:
        return (index - 6, 2)
    if index < 17:
        return (index - 12, 1)
    return (index - 17 + 1, 3)
